src/Storages/StorageKeeperMap.cpp

```diff
@@ -0,0 +1,694 @@
+
+#include <base/types.h>
+
+#include <boost/algorithm/string/classification.hpp>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+    extern const int BAD_ARGUMENTS;
+    extern const int KEEPER_EXCEPTION;
+    extern const int LOGICAL_ERROR;
+    extern const int LIMIT_EXCEEDED;
+}
+
+namespace
+{
+
+std::string_view getBaseName(const std::string_view path)
+{
+    auto last_slash = path.find_last_of('/');
+    if (last_slash == std::string_view::npos)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Failed to get basename of path '{}'", path);
+
+    return path.substr(last_slash + 1);
+}
+
+}
+
+class StorageKeeperMapSink : public SinkToStorage
+{
+    StorageKeeperMap & storage;
+    std::unordered_map<std::string, std::string> new_values;
+    size_t primary_key_pos;
+
+public:
+    StorageKeeperMapSink(StorageKeeperMap & storage_, const StorageMetadataPtr & metadata_snapshot)
+        : SinkToStorage(metadata_snapshot->getSampleBlock()), storage(storage_)
+    {
+        auto primary_key = storage.getPrimaryKey();
+        assert(primary_key.size() == 1);
+        primary_key_pos = getHeader().getPositionByName(primary_key[0]);
+    }
+
+    std::string getName() const override { return "StorageKeeperMapSink"; }
+
+    void consume(Chunk chunk) override
+    {
+        auto rows = chunk.getNumRows();
+        auto block = getHeader().cloneWithColumns(chunk.detachColumns());
+
+        WriteBufferFromOwnString wb_key;
+        WriteBufferFromOwnString wb_value;
+
+        for (size_t i = 0; i < rows; ++i)
+        {
+            wb_key.restart();
+            wb_value.restart();
+
+            size_t idx = 0;
+            for (const auto & elem : block)
+            {
+                elem.type->getDefaultSerialization()->serializeBinary(*elem.column, i, idx == primary_key_pos ? wb_key : wb_value);
+                ++idx;
+            }
+
+            auto key = base64Encode(wb_key.str(), true);
+            new_values[std::move(key)] = std::move(wb_value.str());
+        }
+    }
+
+    void onFinish() override
+    {
+        auto zookeeper = storage.getClient();
+
+        Coordination::Requests requests;
+
+        auto keys_limit = storage.keysLimit();
+
+        size_t current_keys_num = 0;
+        size_t new_keys_num = 0;
+
+        if (keys_limit != 0)
+        {
+            Coordination::Stat root_stat;
+            zookeeper->get(storage.rootKeeperPath(), &root_stat);
+            // exclude metadata node
+            current_keys_num = root_stat.numChildren - 1;
+        }
+
+        std::vector<std::pair<const std::string *, std::future<Coordination::ExistsResponse>>> exist_responses;
+        for (const auto & [key, value] : new_values)
+        {
+            auto path = storage.fullPathForKey(key);
+
+            exist_responses.push_back({&key, zookeeper->asyncExists(path)});
+        }
+
+        for (auto & [key, response] : exist_responses)
```

Comment:
Do we also have a limitations for number of keys per INSERT query and for size of value? I makes sense to add such limitations, because Apache ZooKeeper just closes connection if request size exceeds 1MB (by default). Also the check for `keys_limit` does not guarantee that the limit will not be exceeded in case of multiple parallel INSERTs. It will be much harder to exceed this limit like 10 times if we have some small limit for number of keys per insert.

Reply:
> Also the check for keys_limit does not guarantee that the limit will not be exceeded in case of multiple parallel INSERTs. It will be much harder to exceed this limit like 10 times if we have some small limit for number of keys per insert.

I raised the same issue, but we decided that it's okay to have a soft limit and that we don't care if cases you described happen.

> for size of value

This is interesting, will see what to do for that.

Reply:
> I raised the same issue, but we decided that it's okay to have a soft limit and that we don't care if the cases you described happen.

Ok, please add a comment then